src/chainstate/stacks/index/marf.rs

```diff
@@ -0,0 +1,2205 @@
+        let cur_bhh = storage.tell();
+        if storage.num_blocks() == 0 {
+            // brand new storage
+            trace!("Brand new storage -- start with {:?}", new_bhh);
+            storage.extend(new_bhh)?;
+            let node = TrieNode256::new(&vec![]);
+            let hash = get_node_hash(&node, &vec![]);
+            storage.write_node(&TrieNodeType::Node256(node), hash)
+        }
+        else {
+            // existing storage
+            match storage.open(new_bhh, true) {
+                Ok(_) => {
+                    trace!("Switch to Trie {:?}", new_bhh);
+                    Ok(())
+                }
+                Err(e) => {
+                    match e {
+                        Error::NotFoundError => {
+                            // bring root forward
+                            trace!("Extend {:?} to {:?}", &cur_bhh, new_bhh);
+                            storage.open(&cur_bhh, true)?;
+                            storage.extend(new_bhh)?;
+                            MARF::root_copy(storage, &cur_bhh)?;
+                            storage.open(new_bhh, false)?;
+                            let root_ptr = storage.root_ptr();
+                            fseek(storage, root_ptr)?;
+                            Ok(())
+                        },
+                        _ => {
+                            Err(e)
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /// Walk down this MARF at the given block hash, doing a copy-on-write for intermediate nodes in this block's Trie from any prior Tries.
+    /// s must point to the last filled-in Trie -- i.e. block_hash points to the _new_ Trie that is
+    /// being filled in.
+    fn walk_cow(storage: &mut S, block_hash: &BlockHeaderHash, path: &TriePath) -> Result<TrieCursor, Error> {
+        MARF::extend_trie(storage, block_hash)?;
+
+        let root_ptr = storage.root_ptr();
+        let mut cursor = TrieCursor::new(path, root_ptr);
+
+        // walk to insertion point 
+        let (mut node, _) = Trie::read_root(storage)?;
+        let mut node_ptr = TriePtr::new(0,0,0);
+
+        for _ in 0..(cursor.path.len()+1) {
+            let next_opt = Trie::walk_from(storage, &node, &mut cursor)?;
+            match next_opt {
+                Some((next_node_ptr, next_node, _)) => {
+                    // keep walking
+                    node = next_node;
+                    node_ptr = next_node_ptr;
+                    continue;
+                },
+                None => {
+                    if cursor.div() {
+                        // we're done -- path diverged.  No node-copying can help us.
+                        trace!("Path diverged -- we're done.");
+                        storage.open(block_hash, true)?;
+                        fseek(storage, node_ptr.ptr() as u64)?;
+                        return Ok(cursor);
+                    }
+                    else if cursor.eop() {
```

Comment:
What is the difference between `.eop` and `.eonp` ?

Reply:
`.eop()` == "end of path".  It will return `true` only if the cursor has walked past the last byte of the path.  It will return `false` if there are still bytes to walk in the path.

`.eonp()` == "end of node path".  It will return `true` only if the cursor has walked past the last byte of the compressed path segment within the currently visited node.  It will return `false` if the path being walked and the node's compressed path diverge from one another.

For example, consider the path `00 01 02 03 04 05 06 07 08 09` (paths in the code are 32 bytes, but this will work for our purposes).  Suppose that there's a trie composed of a `Node256` with a child `Node4`, which in turn has a `Leaf`, with the following compressed path segments:

```
Node256[path=00010203] - (04)Node4[path=050607] - (08)Leaf[path=09]=xxxxxxxx
```

So, the `Node256` has the compressed path segment `00 01 02 03`, the `Node4` has the compressed path segment `05 06 07`, and the `Leaf` has the compressed path segment `09`.

Now, suppose we create a `TrieCursor` on the path `00 01 02 03 04 05 ff 07 08 09`.  On the first call to `.walk()`, the cursor will return the pointer for character `04`, and `.eonp()` will be `true` and `.eop()` will be `false`.  This is because the call to `.walk()` will advance the cursor through the `Node256`'s compressed path segment, and reach child `04`.  Since the slot for child `04` is filled, `.walk()` returns a pointer (and not `None`).

On the second call to `.walk()`, the cursor will visit the `Node4` and return `None` once it reaches `ff` in the path.  Then, `.eonp()` will be false because the end of the `Node4`'s path has not been reached (also, `.div()` will return `true` since the act of following the given path `div`erged from all Trie paths encoded in the nodes).  `.eop()` will be `false` because we're not at the end.

Suppose we instead walk the path `00 01 02 03 04 05 06 07 ff 09`.  We'll visit the `Node256` as before, but on the second call to `.walk()`, we'll reach the end of the `Node4`'s path segment and return `None` since it has no child pointer for `ff`.  Here, `.eonp()` will be `true` and `.eop()` will be false.

If we walk the path `00 01 02 03 04 05 06 07 08 ff`, we'll visit the `Node256` on the first call to `.walk()`, the `Node4` on the second call to `.walk()`, and the `Leaf` on the third call to `.walk()`.  Here, `.eonp()` will be `false` (since we did not consume the `Leaf`'s path segment fully) and `.eop()` will be `false`.

Only if we walk the path `00 01 02 03 04 05 06 07 08 09` will `.eop()` be true on the third call to `.walk()`.  After the first two calls to `.walk()`, we'll get back a pointer to another node and `.eonp()` will be `true`.

Does that make sense?

Reply:
Yep, it makes sense. I do think we should try to avoid variable and function names that are long acronyms (eop = is end of path, eonp = is end of node path) and common prefixes (div = diverged or is it divide?)

Reply:
My rationale for this particular name choice was that it closely mirrors the naming conventions in other cursor-like objects.  `.eop()` and `.eonp()` are similar in my mind to `EOF` on files.  Same goes for `.tell()` and `.ntell()` as analogs to `ftell(3)`.  The fact that `.div()` does not take an argument tells me that it cannot possibly mean "divide" -- there's no divisor.  But I'm happy to change them all if you prefer.